app/grandchallenge/algorithms/forms.py

```diff
@@ -622,3 +639,309 @@ class AlgorithmPublishForm(AlgorithmPublishValidation, ModelForm):
+
+    def _build_interfaces(self):
+        remote_interfaces = {
+            interface["slug"]: interface
+            for interface in chain(
+                self.algorithm_serializer.initial_data["inputs"],
+                self.algorithm_serializer.initial_data["outputs"],
+            )
+        }
+
+        self.new_interfaces = []
+        for slug, remote_interface in remote_interfaces.items():
+            try:
+                self._validate_existing_interface(
+                    slug=slug, remote_interface=remote_interface
+                )
+            except ObjectDoesNotExist:
+                # The remote interface does not exist locally, create it
+                self._create_new_interface(
+                    slug=slug, remote_interface=remote_interface
+                )
+
+    def _validate_existing_interface(self, *, remote_interface, slug):
+        serialized_local_interface = ComponentInterfaceSerializer(
+            instance=ComponentInterface.objects.get(slug=slug)
+        )
+
+        for key, value in serialized_local_interface.data.items():
+            # Check all the values match, some are allowed to differ
+            if (
+                key not in {"pk", "description"}
+                and value != remote_interface[key]
+            ):
+                raise ValidationError(
+                    f"Interface {key} does not match for `{slug}`"
+                )
+
+    def _create_new_interface(self, *, remote_interface, slug):
+        new_interface = ComponentInterfaceSerializer(data=remote_interface)
+
+        if not new_interface.is_valid():
+            raise ValidationError(f"New interface '{slug}' is invalid")
+
+        self.new_interfaces.append(new_interface)
+
+    def save(self):
+        self._save_new_interfaces()
+        self._save_new_algorithm()
+        self._save_new_algorithm_image()
+
+    def _save_new_interfaces(self):
+        for interface in self.new_interfaces:
+            interface.save(
+                # The interface kind is a read only display value, this could
+                # be better solved with a custom DRF Field but deadlines...
+                kind=get(
+                    [
+                        c[0]
+                        for c in InterfaceKindChoices.choices
+                        if c[1] == interface.initial_data["kind"]
+                    ]
+                ),
+                store_in_database=False,
+            )
+
+            # Force the given slug to be used
+            interface.instance.slug = interface.initial_data["slug"]
+
+            # Set the store in database correctly, for most interfaces this is
+            # False, then switch it if the super kind is different
+            if interface.initial_data[
+                "super_kind"
+            ] != interface.get_super_kind(obj=interface.instance):
+                interface.instance.store_in_database = True
+            interface.instance.save()
```

Comment:
This makes sense, but if we serialized `store_in_database` on the `ComponentInterfaceSerializer` this would be much more straightforward. Is there a reason why we are not adding that to the serializer?

Reply:
Too many details to expose to external consumers, it's an implementation detail.